### ISSUE TITLE
Suggestion - move control negation lowering from LowerToQIR to its own pass

### ISSUE THREAD

**Issue by amccaskey** [NVIDIA/cuda-quantum#879]
I suggest we remove the negation logic from the op lowering in LowerToQIR and place it in a separate pass that runs before those rewrite patterns, and add it to the pipeline.

**Comment by boschmitt** [comment]
Now that we have changed the IR and I plan to move the decomposition pass from the memory semantics representation to the value semantics one, I believe we can remove the control negation attribute and make negated controls only syntax sugar that will be translated to `quake.x` when building the IR from the AST. (I think this would be a move towards the original behavior, but before changing how the value semantics representation works.)

**Comment by amccaskey** [comment]
What do you mean by syntactic sugar? Could you provide a pseudo mlir snippet showing what would be produced from the AST?

**Comment by schweitzpgi** [comment]
IIRC, it wasn't as simple as mechanically inserting gates before and after. I think it was allowed to negate controls passed to an applicative, e.g. `cudaq::control`, with a callable argument and the callable is pretty free form...

**Comment by boschmitt** [comment]
> What do you mean by syntactic sugar? Could you provide a pseudo mlir snippet showing what would be produced from the AST?

The original behavior was that the presence of a negate control would add a `quake.x` before and after the operation that used it. This is what I mean by syntax sugar: There was not dedicated way of representing negated controls in the IR.

My recollection is that we added a mechanism to represent negated control in the IR, namely, the attribute in the operations because it would greatly facilitate decomposition when done using the memory semantics form of Quake---with recent changes is no longer necessary. However, I would need to look into the point brought up by Eric as it seems I might be missing some other reasoning behind it.

**Comment by amccaskey** [comment]
I see. As I understand it, the only thing we added was the operation attribute to indicate True/False that the control is negated. And the X operations were not injected at the Quake level, but instead were just added when we lowered to LLVM before simulation (maybe I'm wrong... 🤷 ). My proposal for the short term would be to pull that X operation injection out of the LLVM lowering and up to a Quake-level pass.
